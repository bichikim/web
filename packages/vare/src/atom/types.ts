import {UnwrapNestedRefs} from '@winter-love/use'
import {AnyFunction, DropParameters} from '@winter-love/utils'
import {ComputedRef, Ref} from 'vue-demi'

export type AtomRecipe<T extends UnwrapNestedRefs<any>, Args extends any[] = any[]> =
  (state: T, ...args: Args) => unknown

export type AtomGetterRecipe<T extends UnwrapNestedRefs<any>, Return> = (state: T) => Return

export const AtomSymbol = Symbol('atom')
export const ActionWatchSymbol = Symbol('atom action watch')
export const GetterSymbol = Symbol('getter')
export const AtomComputedRefSymbol = Symbol('atom computed ref')
export const ActionSymbol = Symbol('action symbol')

export type GetterRecipe<Return> = (state) => Return

export type GetterRecipeInside<Return> = GetterRecipe<Return> & {
  [GetterSymbol]: true
}

export type AtomType<T> = UnwrapNestedRefs<T> & {
  [ActionWatchSymbol]: any
  [AtomSymbol]: true
}

export type CheckPromise<T extends Promise<unknown> | unknown> =
  T extends Promise<any> ? Promise<boolean> : boolean

export type AtomTypeWithRecipe<
  State,
  Recipe extends AtomRecipe<UnwrapNestedRefs<State>> | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>
  > = AtomType<State> & {
  readonly $: Recipe extends {[GetterSymbol]: true}
    ? (ComputedRef<ReturnType<Recipe>>)
    : ((...args: DropParameters<Recipe>) => unknown)
}

export interface Getter<T> {
  readonly value: T
}

export type AtomReturnTree<T extends Record<string, AnyFunction>, S = any> = {
  [P in keyof T]: T[P] extends {[GetterSymbol]: true}
    ? (Getter<ReturnType<T[P]>>)
    : ((...args: DropParameters<T[P], S>) => ReturnType<T[P]>)
}

export type AtomTypeWithRecipeTree<State extends Record<string, any>,
  Recipe extends AtomRecipe<UnwrapNestedRefs<State>>,
  TreeOptions extends Record<string, Recipe>> = AtomType<State> & {
  readonly $: AtomReturnTree<TreeOptions>
}

export type Atom<State> =
  | AtomType<State>
  //
  | AtomTypeWithRecipe<State, AtomRecipe<UnwrapNestedRefs<State>> | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>>
  //
  | AtomTypeWithRecipeTree<State, AtomRecipe<UnwrapNestedRefs<State>>
  | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>, Record<string, AtomRecipe<UnwrapNestedRefs<State>>
  | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>>>

export type AtomIdentifierName = 'atom'

export type _MayAtomType<State> =
  | State
  | UnwrapNestedRefs<State>
  | AtomType<State>
  | AtomTypeWithRecipe<State, AtomRecipe<UnwrapNestedRefs<State>>
  | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>>
  | Ref<State>

export type MayAtomType<State> =
  | (() => _MayAtomType<State>)
  | _MayAtomType<State>

