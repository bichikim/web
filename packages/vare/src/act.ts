import {
  getGlobalInfo, getIdentifier,
} from 'src/info'
import {
  AnyFunction, FunctionObject,
} from '@winter-love/utils'
import {DropFunctionObject} from 'src/types'
import {
  createGetAtomPrams, createUuid,
} from 'src/utils'
import {shallowRef} from 'vue-demi'
import {
  AnyStateGroup, relateState,
} from 'src/state'

export type ActionRecipe<Args extends any[] = any[], Return = any> = (...args: Args) => Return | Promise<Return>
export type ActionStateRecipe<State = any, Args extends any[] = any[], Return = any> =
  (state: State, ...args: Args) => Return
export type RelatedActionRecipe<State, Args extends any[], Return> =
  (state: State, ...args: Args) => Return | Promise<Return>

export type ActionIdentifierName = 'action'

export const actionName: ActionIdentifierName = 'action'

export type Action<Args extends any[], Return = any> =
  ((...args: Args) => Return | Promise<Return>)

/**
 * check if it is an action
 * only work in development NODE_ENV
 * @param value
 */
export const isAction = (value?: any): value is Action<any> => (getIdentifier(value) === actionName)

const getActPrams = createGetAtomPrams(createUuid('unknown'))

const _act = <Args extends any[], Return> (
  unknown: any,
  mayRecipe: any,
  name?: string,
): Action<Args> => {
  const {
    state, recipe, name: _name,
  } = getActPrams(unknown, mayRecipe, name)
  const watchFlag = shallowRef<any[]>()

  const self: any = (...args: Args): Return | Promise<Return> => {
    const newArgs = state ? [state, ...args] : args

    if (process.env.NODE_ENV === 'development') {
      watchFlag.value = args
    }

    return recipe(...newArgs)
  }

  if (process.env.NODE_ENV === 'development') {
    const info = getGlobalInfo()
    info?.set(self, {
      identifier: actionName,
      name: _name,
      relates: new Set(),
      trigger: watchFlag,
    })

    if (state) {
      relateState(state, self)
    }
  }

  return self
}

const getTreeActPrams = (mayState: any, mayTree: any) => {
  let tree
  let state
  if (mayTree) {
    tree = mayTree
    state = mayState
  } else {
    tree = mayState
  }

  return {
    state,
    tree,
  }
}

const _treeAct = <K extends string, F extends AnyFunction> (
  mayState: any,
  mayTree: Record<K, F>,
): Record<K, (...args: Parameters<F>) => ReturnType<F>> => {
  const {
    tree, state,
  } = getTreeActPrams(mayState, mayTree)
  return Object.keys(tree).reduce((result, name) => {
    const value = tree[name]
    if (state) {
      result[name] = _act(state, value, name)
    } else {
      result[name] = _act(value, name)
    }
    return result
  }, {} as Record<any, any>)
}

export function act<State extends AnyStateGroup, Args extends any[], Return = any>(
  state: State,
  recipe: RelatedActionRecipe<State, Args, Return>,
  name?: string,
): Action<Args>
export function act<Args extends any[], Return> (
  recipe: ActionRecipe<Args, Return>,
  name?: string,
): Action<Args>
export function act<Func extends ActionRecipe, TreeOptions extends Record<string, Func>>(
  tree: TreeOptions
): FunctionObject<TreeOptions>
export function act<
  State extends AnyStateGroup,
  Func extends ActionStateRecipe<State>,
  TreeOptions extends Record<string, Func>
  >(
  state: State,
  tree: TreeOptions,
): DropFunctionObject<TreeOptions, State>
export function act(
  unknown: any,
  mayTree?: any,
  name?: any,
): any {
  if (typeof unknown === 'function' || typeof mayTree === 'function') {
    return _act(unknown, mayTree, name)
  }
  return _treeAct(unknown, mayTree)
}

// todo create actRef
