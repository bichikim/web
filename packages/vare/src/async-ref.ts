import {ref, Ref} from 'vue-demi'
import {freeze} from '@winter-love/utils'

export type AsyncRefRecipe<Args extends any[], Return> = (...args: Args) => Promise<Return> | Return

export interface AsyncRefReturnType<Args extends any[], Return> {
  error: Ref<undefined | any>
  execute(...args: Args): Promise<Return>
  isInProgress: Ref<boolean>
  value: Ref<undefined| Return>
}

export const asyncRef = <Args extends any[], Return>(
  recipe: AsyncRefRecipe<Args, Return>,
): Readonly<AsyncRefReturnType<Args, Return>> => {
  const isInProgress = ref(false)
  const error = ref<undefined | any>()
  const value = ref<Return>()

  const execute = async (...args: Args): Promise<Return> => {
    isInProgress.value = true
    try {
      const result = await recipe(...args)
      value.value = result
      isInProgress.value = false
      return result
    } catch (_error) {
      isInProgress.value = false
      error.value = _error
      throw _error
    }
  }

  return freeze({
    error,
    execute,
    isInProgress,
    value,
  })
}
