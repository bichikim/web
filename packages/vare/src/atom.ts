import {Ref} from 'vue-demi'
import {MayRef, wrapRef} from '@winter-love/use'

export type AtomRecipe<T, Args extends any[]> = (...args: Args) => Promise<T> | T
export type AtomReturnType<T> = Ref<T>
export interface AtomReturnTypeWithRecipe<T, Args extends any[]> extends AtomReturnType<T>{
  act(...args: Args): unknown
}

export type AtomFunctionInitState<T> = () => T

export function atom<T>(initState: MayRef<T>): AtomReturnType<T>
export function atom<T, Args extends any[]>(
  initState: MayRef<T>, recipe?: AtomRecipe<T, Args>): AtomReturnTypeWithRecipe<T, Args>
export function atom<T, Args extends any[]>(initState: MayRef<T>, recipe?: AtomRecipe<T, Args>): any {
  const valueRef = wrapRef(initState)

  if (recipe) {
    return new Proxy(valueRef, {
      get: (target, point) => {
        if (point === 'act') {
          return (...args: Args) => {
            const result = recipe(...args)
            if (result instanceof Promise) {
              result.then((data) => {
                Reflect.set(target, 'value', data)
              })
              return
            }
            Reflect.set(target, 'value', result)
          }
        }
        return Reflect.get(valueRef, 'value')
      },
      set: (target, point, args: Args) => {
        return Reflect.set(target, point, args)
      },
    })
  }

  return valueRef
}
