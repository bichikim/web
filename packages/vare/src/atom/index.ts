import {UnwrapNestedRefs} from '@winter-love/use'
import {computed, shallowRef} from 'vue-demi'
import {getGlobalInfo} from '../info'
import {
  ActionSymbol,
  AtomGetterRecipe,
  AtomIdentifierName,
  AtomRecipe,
  AtomSymbol,
  AtomType,
  AtomTypeWithRecipe,
  AtomTypeWithRecipeTree,
  GetterSymbol,
  MayAtomType,
} from './types'
import {createAction, wrapAtom} from './utils'

export * from './types'
export * from './utils'

export const atomName: AtomIdentifierName = 'atom'

export const createTreeAtom = (reactive, trigger, recipe) => {
  const clonedRecipe = Object.keys(recipe).reduce((result, key) => {
    const value = recipe[key]
    if (value[GetterSymbol]) {
      result[key] = computed(() => value(reactive))
    } else {
      result[key] = value
    }

    return result
  }, {})

  const relates = new Map<string, any>(Object.entries(clonedRecipe))

  const action = new Proxy({}, {
    get: (target, point) => {
      const recipe = clonedRecipe[point]
      if (typeof recipe === 'function') {
        return createAction(trigger, reactive, recipe)
      }
      return recipe
    },
  })

  return {
    atom: new Proxy(reactive, {
      get: (target, point) => {
        switch (point) {
          case '$':
            return action
          default:
            return Reflect.get(target, point, target)
        }
      },
      set: (target, point, value) => {
        return Reflect.set(target, point, value, target)
      },
    }),
    relates,
  }
}

export const createFunctionAtom = (reactive, trigger, recipe) => {
  const clonedRecipe = recipe[GetterSymbol] ? computed(() => recipe(reactive)) : recipe
  const relates = new Map([['default', clonedRecipe]])
  return {
    atom: new Proxy(reactive, {
      get: (target, point) => {
        switch (point) {
          case '$':
            return typeof clonedRecipe === 'function'
              ? createAction(trigger, reactive, recipe)
              : clonedRecipe
          default:
            return Reflect.get(target, point, target)
        }
      },
      set: (target, point, value) => {
        return Reflect.set(target, point, value, target)
      },
    }),
    relates,
  }
}

export const createAtom = (reactive) => {
  return new Proxy(reactive, {
    get: (target, point) => {
      return Reflect.get(target, point, target)
    },
    set: (target, point, value) => {
      return Reflect.set(target, point, value, target)
    },
  })
}

export function atom<State>(initState: MayAtomType<State>): AtomType<State>
export function atom<State extends Record<string, any>,
  Recipe extends AtomRecipe<UnwrapNestedRefs<State>> | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>,
  TreeOptions extends Record<string, Recipe>>(
  initState: MayAtomType<State>,
  recipeTree: TreeOptions,
): AtomTypeWithRecipeTree<State, Recipe, TreeOptions>
export function atom<State,
  Recipe extends AtomRecipe<UnwrapNestedRefs<State>> | AtomGetterRecipe<UnwrapNestedRefs<State>, unknown>,
  >(
  initState: MayAtomType<State>,
  recipe?: Recipe,
): AtomTypeWithRecipe<State, Recipe>
export function atom<T extends Record<string, any>>(
  initState: MayAtomType<T>,
  recipe?: any,
  name?: string,
): any {
  const valueReactive = wrapAtom(initState)
  const trigger = shallowRef<any[]>()
  let relates
  let atom

  // recipe = function
  if (typeof recipe === 'function') {
    const result = createFunctionAtom(valueReactive, trigger, recipe)
    // eslint-disable-next-line prefer-destructuring
    atom = result.atom
    // eslint-disable-next-line prefer-destructuring
    relates = result.relates
    // recipe tree
  } else if (typeof recipe === 'object' && !Array.isArray(recipe)) {
    const result = createTreeAtom(valueReactive, trigger, recipe)
    // eslint-disable-next-line prefer-destructuring
    atom = result.atom
    // eslint-disable-next-line prefer-destructuring
    relates = result.relates
    // recipe none
  } else {
    atom = createAtom(valueReactive)
  }

  // save information for devtool
  if (process.env.NODE_ENV === 'development') {
    const info = getGlobalInfo()

    info?.set(atom, {
      identifier: atomName,
      name,
      relates,
      state: valueReactive,
      trigger,
    })
  }

  // symbol mark
  atom[AtomSymbol] = true
  atom[ActionSymbol] = trigger

  return atom
}

