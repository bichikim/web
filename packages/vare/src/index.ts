import {plugin} from './plugin'
export * from './atom'
export * from './act'
export * from './compute'
export * from './create-devtool'
export * from './info'
export * from './mutate'
export * from './plugin'
export * from './state'
export * from './types'
export * from './utils'
export default plugin
