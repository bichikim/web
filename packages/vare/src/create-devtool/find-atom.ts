import {isAtom} from 'src/atom'
export const findAtom = (target: any, deeps: string[] = []) => {
  const atoms: any[] = []

  if (typeof target === 'object' && !Array.isArray(target)) {
    Object.keys(target).forEach((key) => {
      const value = target[key]
      atoms.push(...findAtom(value, [...deeps, key]))
    })
  }

  if (deeps.length === 0) {
    return atoms
  }

  if (isAtom(target)) {
    atoms.push([deeps.join('.'), target])
    return atoms
  }

  return atoms
}
