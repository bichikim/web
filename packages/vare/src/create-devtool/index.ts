import {DevtoolsPluginApi, setupDevtoolsPlugin, TimelineEvent} from '@vue/devtools-api'
import {drop, isSSR} from '@winter-love/utils'
import {watchTrigger} from 'src/create-devtool/watch-trigger'
import {VareInfo} from 'src/info'
import {App} from 'vue-demi'
import {createStateBases} from './create-state-bases'
import {ApiSetting, CreateDevToolOptions} from './types'
import {watchState} from './watch-state'
import {atomName} from 'src/atom'
import {createInspectorTree} from './create-inspector-tree'

export const DEVTOOL_ID = 'com.npmjs.packages.vare'

// eslint-disable-next-line max-lines-per-function
export const createDevTool = (
  app: App,
  targets: Record<string, any>,
  options: CreateDevToolOptions = {},
) => {
  if (isSSR()) {
    return
  }
  const stateBases = createStateBases(targets)

  const {
    id: inspectorId = 'vare-structure',
    label = 'Vare',
    timeLines = {
      [atomName]: {
        color: 0xF08D49,
        label: 'Vare Atom Actions',
      },
    },
  } = options

  let _api: DevtoolsPluginApi<ApiSetting>

  setupDevtoolsPlugin({
    app,
    id: DEVTOOL_ID,
    label,
    packageName: 'vare',
  }, (api) => {
    _api = api

    api.addInspector({
      icon: 'mediation',
      id: inspectorId,
      label: `${label} Structure`,
      stateFilterPlaceholder: 'Search for state',
      treeFilterPlaceholder: `Search for ${label}`,
    })

    Object.keys(timeLines).forEach((id: string) => {
      const info = timeLines[id]
      api.addTimelineLayer({
        color: info.color,
        id,
        label: info.label,
      })
    })

    api.on.getInspectorTree((payload) => {
      if (payload.app !== app || payload.inspectorId !== inspectorId) {
        return
      }

      const {nodes} = createInspectorTree(targets)
      console.log(nodes)
      payload.rootNodes = nodes
    })

    api.on.getInspectorState((payload) => {
      if (payload.app !== app || payload.inspectorId !== inspectorId) {
        return
      }

      const state = stateBases[payload.nodeId]

      if (state) {

        payload.state = state.refresh()
      }
    })

    api.on.editInspectorState((payload) => {
      if (payload.app !== app || payload.inspectorId !== inspectorId) {
        return
      }

      const state = stateBases[payload.nodeId]

      if (state) {
        const path = drop(payload.path)
        const _state = state.base
        const {value} = payload.state
        payload.set(_state.value, path, value)
      }
    })
  })

  const updateTimeline = (layerId: string, event: Omit<TimelineEvent, 'time'>, all?: boolean) => {
    _api?.addTimelineEvent({
      all,
      event: {
        ...event,
        time: Date.now(),
      },
      layerId,
    })
  }

  const updateTrigger = (info?: VareInfo<string>) => {
    if (!info) {
      return
    }

    updateTimeline(info.identifier, {
      data: {
        type: info.identifier,
      },
    })
  }

  const updateTree = () => {
    _api?.sendInspectorTree(inspectorId)
  }

  const updateState = () => {
    _api?.sendInspectorState(inspectorId)
  }

  watchState(targets, updateState)
  watchTrigger(targets, updateTrigger)

  return {
    updateState,
    updateTimeline,
    updateTree,
  }
}
