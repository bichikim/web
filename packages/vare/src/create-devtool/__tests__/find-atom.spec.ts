import {findAtom} from '../find-atom'
import {atom} from 'src/atom'

describe('find-atom', () => {
  it('should return atoms', () => {
    const result = findAtom(atom({
      foo: atom({
        bar: 'bar',
        john: atom({
          name: 'john',
        }),
      }),
    }))

    expect(result).toMatchSnapshot()
  })
})
