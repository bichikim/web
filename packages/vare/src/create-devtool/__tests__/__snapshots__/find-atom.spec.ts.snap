// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find-atom should return atoms 1`] = `
Array [
  Array [
    "foo.john",
    Object {
      "name": "john",
      Symbol(atom): true,
      Symbol(atom action): undefined,
    },
  ],
  Array [
    "foo",
    Object {
      "bar": "bar",
      "john": Object {
        "name": "john",
        Symbol(atom): true,
        Symbol(atom action): undefined,
      },
      Symbol(atom): true,
      Symbol(atom action): undefined,
    },
  ],
]
`;
